receiver/prometheusreceiver/internal/transaction.go

```diff
@@ -306,38 +311,63 @@ func (t *transaction) getSeriesRef(ls labels.Labels, mtype pmetric.MetricType) u
 
 // getMetrics returns all metrics to the given slice.
 // The only error returned by this function is errNoDataToBuild.
-func (t *transaction) getMetrics(resource pcommon.Resource) (pmetric.Metrics, error) {
+func (t *transaction) getMetrics() (pmetric.Metrics, error) {
 	if len(t.families) == 0 {
 		return pmetric.Metrics{}, errNoDataToBuild
 	}
 
 	md := pmetric.NewMetrics()
-	rms := md.ResourceMetrics().AppendEmpty()
-	resource.CopyTo(rms.Resource())
-
-	for scope, mfs := range t.families {
-		ils := rms.ScopeMetrics().AppendEmpty()
-		// If metrics don't include otel_scope_name or otel_scope_version
-		// labels, use the receiver name and version.
-		if scope == emptyScopeID {
-			ils.Scope().SetName(receiverName)
-			ils.Scope().SetVersion(t.buildInfo.Version)
-		} else {
-			// Otherwise, use the scope that was provided with the metrics.
-			ils.Scope().SetName(scope.name)
-			ils.Scope().SetVersion(scope.version)
-			// If we got an otel_scope_info metric for that scope, get scope
-			// attributes from it.
-			attributes, ok := t.scopeAttributes[scope]
-			if ok {
-				attributes.CopyTo(ils.Scope().Attributes())
-			}
+
+	for rKey, families := range t.families {
+		if len(families) == 0 {
+			continue
+		}
+		resource, ok := t.nodeResources[rKey]
+		if !ok {
+			continue
 		}
-		metrics := ils.Metrics()
-		for _, mf := range mfs {
-			mf.appendMetric(metrics, t.trimSuffixes)
+		rms := md.ResourceMetrics().AppendEmpty()
+		resource.CopyTo(rms.Resource())
+
+		for scope, mfs := range families {
+			ils := rms.ScopeMetrics().AppendEmpty()
+			// If metrics don't include otel_scope_name or otel_scope_version
+			// labels, use the receiver name and version.
+			if scope == emptyScopeID {
+				ils.Scope().SetName(receiverName)
+				ils.Scope().SetVersion(t.buildInfo.Version)
+			} else {
+				// Otherwise, use the scope that was provided with the metrics.
+				ils.Scope().SetName(scope.name)
+				ils.Scope().SetVersion(scope.version)
+				// If we got an otel_scope_info metric for that scope, get scope
+				// attributes from it.
+				if scopeAttributes, ok := t.scopeAttributes[rKey]; ok {
+					if attributes, ok := scopeAttributes[scope]; ok {
+						attributes.CopyTo(ils.Scope().Attributes())
+					}
+				}
+			}
+			metrics := ils.Metrics()
+			for _, mf := range mfs {
+				mf.appendMetric(metrics, t.trimSuffixes)
+			}
 		}
 	}
+	// remove the resource if no metrics were added to avoid returning resources with empty data points
+	md.ResourceMetrics().RemoveIf(func(metrics pmetric.ResourceMetrics) bool {
```

Comment:
is this new behavior that you are fixing, or is this needed to change how resource is identified?

Reply:
This was added to avoid having empty resources without any metrics - previously this was not possible since there were at least the scrape metrics for the resource created in the transaction. However now it could be that a new resource is created based on a new job/instance label, which only contains invalid data (e.g. as in this test case: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3f1c6191271d47cee16970177d2832f3f4684a1c/receiver/prometheusreceiver/internal/transaction_test.go#L338).

However one could argue that in this case it might also make sense to have the empty resourceMetric to indicate the presence of this job/instance pair, but without any data points. This way we would not have to iterate of the resourceMetrics with the `RemoveIf` method - which approach would you say is the better here?

Reply:
Dropping empty sgtm.  Thanks for explaining.